rfcs/splittable-prng.md

```diff
@@ -0,0 +1,103 @@
+
+- If we keep a single global mutable generator state, it needs to be protected by a lock, which makes the PRNG a concurrency bottleneck.
+
+- If we give an independent random generator to each domain, it is unclear how to initialize the state of a new domain. Reusing the state of the parent domain is wrong (the two domain will generate the same random numbers), and inflexible "seeding" policies will be incompatible against some user choices of seed.
+
+Other approaches tend to have dubious randomness properties, and often require tight coupling between the Random library and the multicore runtime, which would increase implementation complexity.
+
+### Splitting saves the day
+
+Some PRNG algorithms provide a "split" operation that returns two pairs of random-generator states, designed to generate independently-looking sequences.
+
+With such "Splittable PRNGs", creating a new domain simply requires splitting the generator state of the parent domain. This approach is efficient (no synchronization of the generator state), has good randomness properties, and it is simple to implement inside the Random module (using runtime-provided thread-local-storage primitives). The decisions of the user in terms of seeding (OS-random or deterministic) are naturally respected, etc. In short: you want a splittable generator for Multicore.
+
+(For a previous discussion of this problem where the suggestion to move to splittable PRNGs emerged, see  https://github.com/ocaml-multicore/ocaml-multicore/pull/582#pullrequestreview-683676282 )
+
+
+## PRINGO generators
+
+The [pringo](https://github.com/xavierleroy/pringo/) library provides two splittable PRNG implementations.
+
+| Name     | numeric type | speed wrt. Random | pure OCaml   | secure            | when to reseed?  |
+|----------|--------------|-------------------|--------------|-------------------|------------------|
+| Random   | int          | =                 | yes          | no                | ?                |
+| SplitMix | int64        | slightly faster   | no (C stubs) | no                | after 2^30 draws | 
+| ChaCha   | int32        | slightly slower   | no (C stubs) | yes (weak crypto) | after 2^64 draws |
+
+To give a sense of the performance difference, on my machine, on a micro-benchmarking drawing numbers in a tight loop (`make benchmarks` from the pringo directory):
+
+- drawing `int` has the same speed for all generators,
+- drawing `int64` takes 55% of the Random time with SplitMix and 93% of the Random time with ChaCha, and
+- drawing `float` values in [0;1] takes 66% of the Random time with SplitMix, and 120% of the Random time with ChaCha.
+
+Those differences are unlikely to be noticeable, as drawing random numbers is a neglectible part of most programs. Some very specific PRNG-bound algorithms (Monte-Carlo simulation, etc.) probably use custom PRNGs anyway.
```

Comment:
I wouldn't be so sure.  QuickCheck-style random property testing can use a lot of pseudo-random numbers.  I'd like to hear from Monte-Carlo specialists re: custom PRNGs.

Reply:
Excellent point! I can try to benchmark QCheck with an alternative PRNG. My guess is that the testing time is dominated by the actual check in user code (running the test), so this would not be noticeable in practice.

I don't know if we have people resembling Monte-Carlo specialists, but my bet would be on @Octachron, with @jhjourdan as the informed hobbyist.

Reply:
> I can try to benchmark QCheck with an alternative PRNG

Or just profile to know how much time is spent in the stdlib PRNG.  Then we can extrapolate from there.

Reply:
As one datapoint Stan (whose one job is Hamiltonian MCMC[0]) uses boost::ecuyer1988 ([1], also known as CombLec88 in eg [2]), using its ability to skip-ahead fast to generate several independant streams ([3] explains why they chose this approach). Unfortunately I have no measure of how much of a typical Stan run the RNG takes (though I suspect little, as Stan spends a lot of time computing likelihoods and derivatives).

[0] sorry if this is not the kind of Monte-Carlo you're thinking of
[1] https://www.boost.org/doc/libs/1_77_0/doc/html/boost_random/reference.html#boost_random.reference.generators
[2] http://videos.rennes.inria.fr/seminaire_Irisa/lecuyer/rngx.pdf
[3] https://github.com/stan-dev/stan/pull/3028#issuecomment-804411734

Reply:
I would expect Markov Chain Monte-Carlo simulation to tune their PRNG since those require to generate a lot of samples from a simple distribution fast to achieve thermalization. I will try to find some of my sampling experience from my PhD to do more performance analysis (but from my memory using efficient algorithm for sampling non-uniform distribution had a larger impact than the PRNG itself).